### ISSUE TITLE
Live migration fail with "Error collecting checkpoint log file"

### ISSUE THREAD

**Issue by fluboi** [lxc/lxd#2466]
# Required information

 * Distribution: Ubuntu 16.04 
 * Kernel version: 4.4.0-38-generic
 * LXC version: 2.0.4
 * LXD version: 2.0.4
 * Storage backend in use: zfs

# Issue description

Live migration end up with the following error:
`error: Error transferring container data: migration failed 1`

Log file /var/log/lxd/lxd.log shows : 
`t=2016-10-06T17:35:17+0100 lvl=eror msg="Error collecting checkpoint log file" err="open /tmp/lxd_checkpoint_986842051/dump.log: no such file or directory"`

/var/log/lxd/lxd.log: https://gist.github.com/fluboi/e2a84fda21c64f524fc0a7c30d2aa8a3
/var/log/lxd/ct1/lxc.log: https://gist.github.com/fluboi/aa0d72c824f8646327e9729ce3c42448

# Steps to reproduce
Basic fresh install with 2 Ubuntu 16.04. "Normal" migration is working properly. (by stopping CT before)

**Comment by stgraber** [comment]
Sounds like criu failed before it even got to write to any log files. @tych0

**Comment by tych0** [comment]
What's your criu version? (`criu --version` should suffice) I thought this usually meant criu didn't understand the arguments, but somehow the arguments aren't listed in your log, which leads me to suspect something is very wrong :(

**Comment by fluboi** [comment]
I'm using criu `Version: 2.0`. (Ubuntu 16.04, repository package)

**Comment by tych0** [comment]
Ok, and I just looked and 2.0.4 should emit the flags it passes to criu, so now I'm confused. Is the above the paste of the whole log?

**Comment by fluboi** [comment]
Yes, I've paste the whole logs created between the `lxc move`command, and the fail  `migration failed 1`

**Comment by fluboi** [comment]
On the remote host (the one that should receive the CT), I also have that lxd log : `lvl=eror msg="Error during migration sink" err="migration failed 1\n"` ... but I'm not sure it helps a lot..

**Comment by tych0** [comment]
I guess I'm not sure what to tell you, other than perhaps try to upgrade to the latest CRIU? I don't see where in the code criu or the liblxc child would exit 2, and I'm not sure why you're not getting the flags that were actually passed to criu.

What's the output of `cat /proc/$lxd_pid/maps`? Do you have a 1.1 liblxc loaded or something?

**Comment by fluboi** [comment]
```
 sudo cat /proc/2429/maps | grep liblxc
7fe854067000-7fe8540f1000 r-xp 00000000 fc:00 394234                     /usr/lib/x86_64-linux-gnu/liblxc.so.1.2.0
7fe8540f1000-7fe8542f1000 ---p 0008a000 fc:00 394234                     /usr/lib/x86_64-linux-gnu/liblxc.so.1.2.0
7fe8542f1000-7fe8542f3000 r--p 0008a000 fc:00 394234                     /usr/lib/x86_64-linux-gnu/liblxc.so.1.2.0
7fe8542f3000-7fe8542f4000 rw-p 0008c000 fc:00 394234                     /usr/lib/x86_64-linux-gnu/liblxc.so.1.2.0
```
Don't know how to read that. It should not be 2.0.X ? Or liblxc version is not linked to LXC version ?

**Comment by tych0** [comment]
No, that's right. hm.

**Comment by tych0** [comment]
Do you have some other executable named `criu` in your path somewhere maybe? `which -a criu`?

**Comment by fluboi** [comment]
I've upgrade criu to 2.6. And... it still not working, but I have a totally different error message. Should I go back to 2.0 version, open a new issue for that :

CLI:
```
error: Error transferring container data: migration dump failed
(00.922683) Error (criu/action-scripts.c:116): One of more action scripts failed
(00.922715) Error (criu/cr-dump.c:1654): Pre dump script failed with 32512!
(00.923072) Error (criu/ptrace.c:54): Unable to detach from 2726: No such process
(01.049008) Error (criu/stats.c:73): BUG at criu/stats.c:73
```
lxc.log:
```
            lxc 20161006214726.466 INFO     lxc_criu - criu.c:exec_criu:419 - execing: /usr/sbin/criu dump --tcp-established --file-locks --link-remap --manage-cgroups=full --ext-mount-map auto --enable-external-sharing --enable-external-masters --enable-fs hugetlbfs --enable-fs tracefs -D /tmp/lxd_checkpoint_974840257 -o /tmp/lxd_checkpoint_974840257/dump.log -vvvvvv --action-script � -t 2726 --freeze-cgroup /sys/fs/cgroup/freezer///lxc/ct1 --ext-mount-map /dev/console:console --external tty[8801:e] --force-irmap --ghost-limit 859539989344 
            lxc 20161006214726.469 DEBUG    lxc_commands - commands.c:lxc_cmd_handler:893 - peer has disconnected
            lxc 20161006214729.320 ERROR    lxc_criu - criu.c:do_dump:920 - dump signaled with 6
```

Maybe the way I've install criu 2.6 is not corect (downloading the .deb package, and installing it using dpkg -i)

Currently `which -a criu` only return `/usr/sbin/criu`. (I haven't test before the upgrade)

**Comment by tych0** [comment]
Heh, this bug should be fixed too, actually. It could be one of two things, one is fixed in 2.0.4 and another is fixed in 2.0.5. How did you get your liblxc and LXD? From the archive?

**Comment by fluboi** [comment]
Exactly what I'v done to upgrade criu:
```
wget https://launchpad.net/ubuntu/+source/criu/2.6-1ubuntu1/+build/10991540/+files/criu_2.6-1ubuntu1_amd64.deb
dpkg -i criu_2.6-1ubuntu1_amd64.deb
```
(Also rebooted after that)

My initial install was just as simple as :
`sudo apt install lxd  zfsutils-linux zfs-dkms bridge-utils criu` in a clean (just installed) Ubuntu server 16.04

I'll try (tomorrow or this week-end) to re-do a fresh install, but I'm pretty sure I'll be facing the same bug as there is nothing eccentric in my current install ! :)

**Comment by fluboi** [comment]
But I'm currently still in LXC & LXD Version 2.0.4 accordind to `lxc info`. I've just upgraded criu.

**Comment by tych0** [comment]
Right, I think this should be fixed in 2.0.5, which should hit the archive soon. You can use the stable PPA in the meantime, though https://launchpad.net/~ubuntu-lxc/+archive/ubuntu/stable

Can you try and upgrade liblxc, do a restart of LXD, and then do a checkpoint of a container?

**Comment by stgraber** [comment]
Hmm, be very careful when using that PPA. For LXD, it will upgrade you from 2.0.4 to 2.4.1. Once that's done, you cannot go back as we don't support database downgrades!

So if you don't mind running the latest feature release and get all the new features every month or so, go for it. But if you want to stick to the production grade release, you should wait for 2.0.5 to make its way through updates rather than upgrade to 2.4.1.

**Comment by fluboi** [comment]
So if I understand well, "stable release" is not really ready for production environments ? Sounds not natural for me as a Debian guy ! 

When you say :
> 2.0.5, which should hit the archive soon
What means "soon" ? days ? weeks ? months ? I need to setup a production environment... soon !

I will test the PPA "stable" in my testing lab, but for the production I would like something... stable ! It would be great if live migration works in the production environment, but if not (and effectively corrected in a near future), it's not dramatic. 

Thanks a lot for your help.

**Comment by stgraber** [comment]
So for all the LXC products (LXC, LXCFS and LXD), we publish normal "stable" releases at various intervals, which introduce new features and may change command line arguments and user experience. We also publish LTS releases of those 3 projects every two years, those LTS releases then get bugfix and security updates but no new feature or user experience change.

So both LXD 2.0.5 and LXD 2.4 are stable. It's just that support for LXD 2.4 will only last a few months before we tell people to move to LXD 2.5 or whatever's current while support for LXD 2.0.x will last 5 years.

As for the PPAs, we have ppa:ubuntu-lxc/stable which is the latest release (as opposed to pre-release or git snapshots) of all products and then we have ppa:ubuntu-lxc/lts which is the latest LTS release of all products.

Anyway, I was travelling this week (still am, sending this from over the atlantic) and only had time to push the most urgent releases to the Ubuntu archive (there was a deadline for Ubuntu 16.10 on Thursday) so that was LXC 2.0.5, LXCFS 2.0.4 and LXD 2.4.1.

Next up is pushing LXC 2.0.5, LXCFS 2.0.4 and LXD 2.0.5 to xenial-proposed which I'll most likely be doing on Monday, after review by the SRU team, the packages will go in proposed for 7 days of testing, if all goes well, they'll then hit xenial-updates. So the most optimistic ETA for 2.0.5 in the Ubuntu 16.04 archive is the 17th of October.

Once those hit xenial-updates, we'll then update the backports we have in trusty-backports for those still on Ubuntu 14.04.

**Comment by fluboi** [comment]
Ok thanks for your clear answer. 
And about criu versions ? Is it totally independent ? Is it possible to do criu downgrades ? (seems yes, but if you can confirm...)

**Comment by stgraber** [comment]
Yeah, CRIU is independent. LXC just calls out to the criu binary, so as long as they don't remove options in a new release you'll be fine. We have detection code for the criu version to drop some arguments when running older releases.

**Comment by fluboi** [comment]
Ok. Thanks. 
I've tried to reproduce the original bug (about checkpoint log file), from clean install. 
You can find a trace of what I've done [here](https://gist.githubusercontent.com/fluboi/fc3be7295cf4d662c39026e1f0fea270/raw/5fa2e00529d5c8265f00837760d526b19e8a1762/debug).

In short: I'm not able to reproduce the same bug. But I'm facing something that look like more the one already underline in [that previous message](https://github.com/lxc/lxd/issues/2466#issuecomment-252085880).

By then adding `ppa:ubuntu-lxc/stable`, update/upgrade/restart LXD, I confirm that the bug disappear and that I'm able to live migrate containers. 

Packages upgraded with ppa add:
```
liblxc1/xenial 2.0.5-0ubuntu1~ubuntu16.04.1~ppa1 amd64 [upgradable from: 2.0.4-0ubuntu1~ubuntu16.04.2]
libseccomp2/xenial 2.3.1-2ubuntu2~ubuntu16.04.1~ppa1 amd64 [upgradable from: 2.2.3-3ubuntu3]
lxc-common/xenial 2.0.5-0ubuntu1~ubuntu16.04.1~ppa1 amd64 [upgradable from: 2.0.4-0ubuntu1~ubuntu16.04.2]
lxcfs/xenial 2.0.4-0ubuntu1~ubuntu16.04.1~ppa1 amd64 [upgradable from: 2.0.3-0ubuntu1~ubuntu16.04.1]
lxd/xenial 2.4.1-0ubuntu1~ubuntu16.04.1~ppa1 amd64 [upgradable from: 2.0.4-0ubuntu1~ubuntu16.04.1]
lxd-client/xenial 2.4.1-0ubuntu1~ubuntu16.04.1~ppa1 amd64 [upgradable from: 2.0.4-0ubuntu1~ubuntu16.04.1]
```
Do you confirm that this specific bug is corrected in the coming LXD 2.0.5 ? :)

**Comment by stgraber** [comment]
All fixes related to migration from master were included, so yeah, the commit is in 2.0.5

**Comment by stgraber** [comment]
Cool, so sounds like the newer liblxc and lxd does fix the issue for you, so closing this issue as it's fixed upstream. I'm hoping we won't get delayed in pushing LXC and LXD 2.0.5 to Ubuntu 16.04 so you've got this fixed soon!
